src/Controller/CRUDController.php

```diff
@@ -524,7 +523,60 @@ public function batchAction(Request $request): Response
             return $this->redirectToList();
         }
 
-        return $this->$finalAction($query, $forwardedRequest);
+        return call_user_func($this->getBatchActionExecutable($action), $query, $forwardedRequest);
+    }
+
+    final protected function canBatchActionBeExecuted(string $action): bool
+    {
+        $batchActions = $this->admin->getBatchActions();
+        if (!\array_key_exists($action, $batchActions)) {
+            throw new \RuntimeException(sprintf('The `%s` batch action is not defined', $action));
+        }
+
+        $controller = $batchActions[$action]['controller'] ?? null;
+
+        if ($controller) {
+            $request = new Request();
+            $request->attributes->set('_controller', $controller);
+            try {
+                return false !== $this->container->get('controller_resolver')->getController($request);
+            } catch (\Throwable $error) {
+                return false;
+            }
+        } else {
+            $camelizedAction = InflectorFactory::create()->build()->classify($action);
+            $finalAction = sprintf('batchAction%s', $camelizedAction);
+
+            if (method_exists($this, $finalAction)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    final protected function getBatchActionExecutable(string $action): callable
+    {
+        if (!$this->canBatchActionBeExecuted($action)) {
+            throw new \LogicException(sprintf('Batch action `%s` cannot be executed. You must call `%s::canBatchActionBeExecuted` first.', $action, static::class));
+        }
```

Comment:
This means the canBatchActionBeExecuted method is executed twice... :s

Reply:
Yes it is. I was going to put the method private, if we do we could not call it again since the flow we have will make sure it cannot happen. If we keep it protected I think it's better to have the check. I'll put it private for now and wait if we have a need to put in protected at some point.